src/subscription.rs

```diff
@@ -284,37 +285,66 @@ pub struct SubscriptionFuture<TRes: Debug + PartialEq> {
-                return Poll::Ready(Err(Cancelled));
-            }
-        };
-
-        match subscription {
-            Subscription::Cancelled => Poll::Ready(Err(Cancelled)),
-            Subscription::Pending { ref mut waker, .. } => {
-                *waker = Some(context.waker().clone());
-                task::block_on(async { self.subscriptions.lock().await })
-                    .get_mut(&self.kind)
-                    .and_then(|subs| subs.insert(self.id, subscription));
-                Poll::Pending
+    fn poll(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<Self::Output> {
+        use std::collections::hash_map::Entry::*;
+
+        // FIXME: using task::block_on ever is quite unfortunate. alternatives which have been
+        // discussed:
+        //
+        // - going back to std::sync::Mutex
+        // - using a state machine
+        //
+        // std::sync::Mutex might be ok here as long as we don't really need to await after
+        // acquiring. implementing the state machine manually might not be possible as all mutexes
+        // lock futures seem to need a borrow, however using async fn does not allow implementing
+        // Drop.
+        let mut subscriptions = task::block_on(async { self.subscriptions.lock().await });
+
+        if let Some(related_subs) = subscriptions.get_mut(&self.kind) {
+            let (became_empty, ret) = match related_subs.entry(self.id) {
+                // there were no related subs, it can only mean cancellation or polling after
+                // Poll::Ready
+                Vacant(_) => return Poll::Ready(Err(Cancelled)),
+                Occupied(mut oe) => {
+                    let unwrapped = match oe.get_mut() {
+                        Subscription::Pending { ref mut waker, .. } => {
+                            // waker may have changed since the last time
+                            *waker = Some(context.waker().clone());
+                            return Poll::Pending;
+                        }
+                        Subscription::Cancelled => {
+                            oe.remove();
+                            Err(Cancelled)
+                        }
+                        _ => match oe.remove() {
+                            Subscription::Ready(result) => Ok(result),
+                            _ => unreachable!("already matched"),
+                        },
+                    };
+
+                    (related_subs.is_empty(), unwrapped)
+                }
+            };
+
+            if became_empty {
+                // early cleanup if possible for cancelled and ready. the pending variant has the
+                // chance of sending out the cancellation message so it cannot be treated here
+                subscriptions.remove(&self.kind);
             }
-            Subscription::Ready(result) => Poll::Ready(Ok(result)),
+
+            // need to drop manually to aid borrowck at least in 1.45
+            drop(subscriptions);
```

Comment:
an extra scope might work too, but this one is less whitespace nesting, so I guess it's fine too :+1:

Reply:
Just tested, this would also work. My posted version spares the reader of one of my famous and loved `let (_, ret) = { ... };` idioms (the `ret` binding) :)